core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -808,6 +815,7 @@ private[spark] class TaskSchedulerImpl(
       // happen below in the rootPool.executorLost() call.
       taskIds.foreach(cleanupTaskState)
     }
+    executorIdToCores.remove(executorId).foreach(totalSlots -= _ / CPUS_PER_TASK)
```

Comment:
I think that rather than calling `rootPool.updateAvailableSlots()` on every `resourceOffer`, it would be more efficient to just call it when the total number has changed -- here and in the `resourceOffer` branch where its updated.  Might not matter much, hopefully `updateAvailableSlots()` isn't too expensive, but still I think that would be better.

Reply:
updateAvailableSlots is dependent on both the total number of slots + what tasksets/pools are present. Will take a look to see how feasible doing the update incrementally is.